internal/telemetry/client.go

```diff
@@ -0,0 +1,407 @@
+		// See https://golang.org/pkg/net/http/#DefaultTransport .
+		Transport: &http.Transport{
+			Proxy: http.ProxyFromEnvironment,
+			DialContext: (&net.Dialer{
+				Timeout:   30 * time.Second,
+				KeepAlive: 30 * time.Second,
+				DualStack: true,
+			}).DialContext,
+			MaxIdleConns:          100,
+			IdleConnTimeout:       90 * time.Second,
+			TLSHandshakeTimeout:   10 * time.Second,
+			ExpectContinueTimeout: 1 * time.Second,
+		},
+		// XXX: Should the timeout be shorter than 5 seconds? Telemetry should
+		// not hold up the application
+		Timeout: 5 * time.Second,
+	}
+	// TODO: Default telemetry URL?
+)
+
+// Client buffers and sends telemetry messages to Datadog (possibly through an
+// agent). Client.Start should be called before any other methods.
+//
+// Client is safe to use from multiple goroutines concurrently. The client will
+// send all telemetry requests in the background, in order to avoid blocking the
+// caller since telemetry should not disrupt an application. Metrics are
+// aggregated by the Client.
+type Client struct {
+	// URL for the Datadog agent or Datadog telemetry endpoint
+	URL string
+	// APIKey should be supplied if the endpoint is not a Datadog agent,
+	// i.e. you are sending telemetry directly to Datadog
+	APIKey string
+	// How often to send batched requests. Defaults to 60s
+	SubmissionInterval time.Duration
+
+	// e.g. "tracers", "profilers", "appsec"
+	Namespace string
+
+	// App-specific information
+	Service string
+	Env     string
+	Version string
+
+	// Determines whether telemetry should actually run.
+	// Defaults to false, but will be overridden by the environment variable
+	// DD_INSTRUMENTATION_TELEMETRY_ENABLED=0
+	Disabled bool
+
+	// Optional destination to record submission-related logging events
+	Logger interface {
+		Printf(msg string, args ...interface{})
+	}
+
+	// mu guards all of the following fields
+	mu sync.Mutex
+	// started is true in between when Start() returns and the next call to
+	// Stop()
+	started bool
+	// seqID is a sequence number used to order telemetry messages by
+	// the back end.
+	seqID int64
+	// t is used to schedule flushing outstanding messages
+	t *time.Timer
+	// requests hold all messages which don't need to be immediately sent
+	requests []*Request
+	// metrics holds un-sent metrics that will be aggregated the next time
+	// metrics are sent
+	metrics    map[string]*metric
+	newMetrics bool
+}
+
+func (c *Client) log(msg string, args ...interface{}) {
+	if c.Logger == nil {
+		return
+	}
+	c.Logger.Printf(msg, args...)
+}
+
+// Start registers that the app has begun running with the given integrations
+// and configuration.
+func (c *Client) Start(integrations []Integration, configuration []Configuration) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	if c.Disabled || os.Getenv("DD_INSTRUMENTATION_TELEMETRY_ENABLED") == "0" {
+		return
+	}
+	if c.started {
+		return
+	}
+	defer func() { c.started = true }()
```

Comment:
NIT: Not sure I understand the need for this defer vs just setting `c.started = true` here directly.

Reply:
I went through a few iterations of the design and I think this was left-over from doing something kind of like `sync.Once` (where you have a quick atomic check if initialization is done and then a slower path for actually doing initialization) But that ended up feeling too complicated. I agree in the current implementation defer is not needed.